2D Elastoplasticity FEM

rishabharora236-cell opened the pull request:
This update introduces 2D J2 plasticity implementation for quadrilateral meshes.

Key Changes:

New Files:

Added a new example script implementing J2 plasticity for a 2D quad mesh.

Added a new loss function module for element-level computation (currently supports only quad meshes).

Modified Files:

FEM utilities: Updated to include computations at the Gauss point level for elastoplastic behavior.

Solver module: Added a new function that supports solving with history-dependent parameters.

fe_loss.py: Adjusted existing functions to handle optional history parameters without breaking backward compatibility. This may appear a bit messy since I prioritized maintaining compatibility with existing functions rather than introducing entirely new ones. A cleaner refactor can also be done.

RezaNajian commented:
@kianoosh1989 Thank you for the feature! This branch has several conflicts with the main branch due to some fundamental changes that are missing in this PR. Could you please merge the main branch into this one and make any necessary adjustments? Once that’s done, we can proceed with the review and merge.

kianoosh1989 commented:
Dear Dr. Najian, 

I updated this PR regarding your last comments. In addition, stress function was added.

Best

RezaNajian committed:
transfer changes from original branch

RezaNajian committed:
rewrite all f-strings safely

RezaNajian commented:
@rishabharora236-cell 

Thanks a lot for your work on introducing the **2D elastoplasticity (J2) implementation** — this is a great addition to the framework and a significant step forward. I went through the code in detail and refactored/cleaned several parts to improve structure, readability, and maintainability.

All relevant element-level and history-dependent computations have now been moved into:

* `mechanical_elastoplasticity.py`
* `fe_nonlinear_residual_based_solver_with_history_update.py`

The example script has also been moved into a dedicated **`elastoplasticity/`** folder for better project organization and overall cleanliness.

### ✔ Follow-up actions

Please take a look at **`mechanical_2d_elastoplasticity.py`** and update the code **after line 98** to align with the new structure.

Additionally, it would be great if you could address the following items:

1. **Add a Unit Test**
   Please provide a unit test for the newly introduced `mechanical_elastoplasticity` loss function so we can properly validate behavior and avoid regressions.

2. **Improve Example Readability**
   Please reorganize and polish `mechanical_2d_elastoplasticity.py` to make it more readable and structured, especially for new users exploring elastoplastic simulations.

3. **Extend to 3D and Additional Element Types**
   As a next step, please consider extending the implementation to:

   * 3D elements
   * Other element types (triangles, hexahedra, etc.)

   The structure in `mechanical_elastoplasticity.py` was designed to allow these extensions naturally.

---

Thanks again for your contribution — your work is helping shape a robust and extensible elastoplastic FEM module! Let me know if you need any support with the follow-up tasks.

rishabharora236-cell commented:
I've reviewed the refactored code—thanks for cleaning that up.

Regarding fol/tools/fem_utilities.py: I noticed the specific material model logic was removed. I plan to re-introduce a dedicated material model entry there (similar to the existing Neo-Hooke implementation). I'll include this in my next commit.

Please do let me know, if you have any suggestion related to this or if I am missing something.

RezaNajian commented:
> I've reviewed the refactored code—thanks for cleaning that up.
> 
> Regarding fol/tools/fem_utilities.py: I noticed the specific material model logic was removed. I plan to re-introduce a dedicated material model entry there (similar to the existing Neo-Hooke implementation). I'll include this in my next commit.
> 
> Please do let me know, if you have any suggestion related to this or if I am missing something.

@rishabharora236-cell 
If the material model logic was removed, that must have been by mistake — thanks for pointing it out.
Before reintroducing it, I would actually recommend restructuring the material modeling part so that it follows a clean, modular FEM/computational-mechanics design.

Specifically:

Create a dedicated folder for constitutive (material) models, e.g. constitutive_material_models/.

Add a BaseConstitutiveModel class defining the common interface.

Move each implemented model (e.g., Neo-Hookean) into its own derived class.

Update the loss functions accordingly to use this class-based structure.

This keeps fem_utilities focused on FEM operations while making constitutive laws modular and extendable.

Let me know if you need help with the restructuring or want to discuss the design — happy to jump in!

Commit:
Modular FEM + 2D and 3D elastoplasticity with unit test

Commit:
Saint_Venant_Material_Model

rishabharora236-cell commented:
Dear Dr. Najian,

Thanks for the detailed feedback! I have updated the code to align with the modular structure you suggested. Here is a summary of the changes:

- Refactored Material Models: I created a constitutive_material_models/ directory and implemented the BaseConstitutiveModel class. The Neo-Hookean and Elastoplastic models now inherit from this base class, and the loss functions have been updated to use this new interface.

- Unit Tests: I added a unit test for the mechanical_elastoplasticity loss function .

- Code Readability: I reorganized mechanical_2d_elastoplasticity.py and renamed several variables to make the logic clearer for new users.

- 3D Extension: I have added an example for 3D elements as requested.

- Saint Venant Model: I have also implemented the Saint Venant material class within the new structure. However, I haven't integrated it into a specific example script yet (as I wasn't sure which example is using this material model). It is currently available for import.

Please do let me know where can I improve further and Thanks.

Commit:
Improved_constitutive_material_model

RezaNajian commented on fol/constitutive_material_models/base.py:
```diff
@@ -0,0 +1,42 @@
+
+
+
+class PlasticityModel(BaseConstitutiveModel):
```
When you get a chance, could you move the PlasticityModel class definition to plasticity.py? That way we can keep the base class purely abstract in one place.
It would also be great if you could add your authorship information at the top of the file so we have proper attribution.

RezaNajian commented on fol/constitutive_material_models/neo_hooke.py:
```diff
@@ -0,0 +1,207 @@
+    """
+    Material model.
+    """
+    @partial(jit, static_argnums=(0,))
```
could you remove the jit decorators from these functions? Since they’re already being called from higher-level functions that are jitted (e.g. ComputeJacobianMatrixAndResidualVector), jitting the nested functions adds overhead rather than improving performance.

RezaNajian commented on fol/constitutive_material_models/plasticity.py:
```diff
@@ -0,0 +1,304 @@
+"""
+J2 (von Mises) plasticity with isotropic hardening.
+
```
if this implementation is based on a specific reference or paper, please mention it so we can include a proper citation for future readers.

RezaNajian commented on fol/constitutive_material_models/solvers.py:
```diff
@@ -0,0 +1,118 @@
+"""
```
I’d suggest moving this implementation to utils.py so we can keep the structure more focused on material models. Since this Newton solver is only used by material models like plasticity, grouping it in utils.py should help keep things organized and easier to navigate.

RezaNajian commented on fol/constitutive_material_models/saint_venant.py:
```diff
@@ -0,0 +1,69 @@
+    Material model.
+    """
+    @partial(jit, static_argnums=(0,))
+    def evaluate(self, F, lambda_, mu, *args, **keyargs):
```
It looks like *args and **kwargs aren’t being used here. If there isn’t a specific reason to keep them, could you remove them to simplify the interface?

RezaNajian commented on fol/constitutive_material_models/saint_venant.py:
```diff
@@ -0,0 +1,69 @@
+    Material model.
+    """
+    @partial(jit, static_argnums=(0,))
+    def evaluate(self, E_mat, lambda_, mu, *args, **keyargs):
```
the same here

RezaNajian commented on fol/loss_functions/mechanical_elastoplasticity.py:
```diff
@@ -0,0 +1,251 @@
+            hardening_exponent=mat_dict["iso_hardening_param_2"] # Use .get() for safety
+        )
+
+    @partial(jit, static_argnums=(0,))
```
You can remove the jit decorations entirely, except for the one on ComputeJacobianMatrixAndResidualVector, which is the function actually called inside the Newton loop. When the high-level function is jitted, there’s no need to jit the lower-level functions—doing so generally adds overhead and also makes debugging more difficult.

RezaNajian commented on fol/tools/fem_utilities.py:
could you resolve the conflicts with the main branch so this can be merged smoothly.

RezaNajian reviewed (changes requested):
@rishabharora236-cell 
Thanks a lot for the thorough updates and for addressing the structural feedback so carefully — I really appreciate the effort you put into reorganizing the material models and improving clarity across the modules.

I went through the changes, and overall everything looks solid. There are just a few minor adjustments I commented on that I’d appreciate if you could apply where possible. Once those are in place, we should be ready to merge.

Thanks again for the great work and the clear summary of your updates — it’s coming together very nicely!

Reza

Commit:
Updates_on_comments

rishabharora236-cell commented:
Dear Dr. Najian,

I have resolved issues mentioned in your comments.

Thanks and Regards
Rishabh

RezaNajian committed:
Merge branch 'main' into feat/2D_Elastoplasticity_FEM

RezaNajian committed:
fix SaintVenan import issue
